docs/RingBuffer.md

```diff
@@ -0,0 +1,304 @@
+ *
+ * Wakeup options (flags):
+ * - 0 (auto/default): Notify if consumer has caught up.
+ * - BPF_RB_FORCE_WAKEUP - Always notify consumer.
+ * - BPF_RB_NO_WAKEUP - Never notify consumer.
+ *
+ */
+ebpf_result_t
+ebpf_ring_buffer_output(ebpf_ring_buffer_t* ring, uint8_t* data, size_t length, size_t flags)
+```
+
+### Updated supported libbpf functions
+
+```c
+
+struct ring_buffer;
+
+typedef int (*ring_buffer_sample_fn)(void *ctx, void *data, size_t size);
+
+struct ring_buffer_opts {
+	size_t sz; /* size of this struct, for forward/backward compatiblity */
+};
+
+#define ring_buffer_opts__last_field sz
+
+/**
+ * @brief Creates a new ring buffer manager.
+ *
+ * @param[in] map_fd File descriptor to ring buffer map.
+ * @param[in] sample_cb Pointer to ring buffer notification callback function.
+ * @param[in] ctx Pointer to sample_cb callback function context.
+ * @param[in] opts Ring buffer options.
+ *
+ * @returns Pointer to ring buffer manager.
+ */
+struct ring_buffer *
+ring_buffer__new(int map_fd, ring_buffer_sample_fn sample_cb, void *ctx,
+		 const struct ring_buffer_opts *opts);
+
+/**
+ * @brief Add extra RINGBUF maps to this ring buffer manager
+ *
+ * @param[in] rb Pointer to ring buffer manager.
+ * @param[in] map_fd File descriptor to ring buffer map.
+ * @param[in] sample_cb Pointer to ring buffer notification callback function.
+ * @param[in] ctx Pointer to sample_cb callback function context.
+ */
+int ring_buffer__add(struct ring_buffer *rb, int map_fd,
+		     ring_buffer_sample_fn sample_cb, void *ctx)
+
+/**
+ * @brief Frees a ring buffer manager.
+ *
+ * @param[in] rb Pointer to ring buffer manager to be freed.
+ *
+ */
+void ring_buffer__free(struct ring_buffer *rb);
+
+/**
+ * @brief poll ringbuf for new data
+ * Poll for available data and consume records, if any are available.
+ * Returns number of records consumed (or INT_MAX, whichever is less), or
+ * negative number, if any of the registered callbacks returned error.
+ *
+ * @param[in] rb Pointer to ring buffer manager.
+ * @param[in] timeout_ms maximum time to wait for (in milliseconds).
+ *
+ */
+int ring_buffer__poll(struct ring_buffer *rb, int timeout_ms);
+
+/**
+ * @brief catch consumer up to producer by invoking the callback for every available record
+ * Consume available ring buffer(s) data without event polling.
+ * Returns number of records consumed across all registered ring buffers (or
+ * INT_MAX, whichever is less), or negative number if any of the callbacks
+ * return error.
+ *
+ * @param[in] rb Pointer to ring buffer manager.
+ */
+int ring_buffer__consume(struct ring_buffer *rb);
+```
+
+### New ebpf APIs
+
+```c
+/**
+ * get pointers to mapped producer and consumer pages
+ *
+ * @param[out] producer pointer* to start of read-only mapped producer pages
+ * @param[out] consumer pointer* to start of read-write mapped consumer page
+ */
+int ebpf_ring_buffer_get_buffer(const ebpf_map_t* map, void **producer, void **consumer);
```

Comment:
To resolve #3914 I'll be updating this to return a handle the user can use to release the mapped memory. This will probably be by using a section object to map views of the producer and consumer pages into memory.

Reply:
If we map views of sections into user space the user process can unmap the sections using the producer/consumer addresses, so the existing signature supports that.